### ISSUE TITLE
 Update the `socket` source to instrument open connection times as a gauge 

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#3518]
Similar to https://github.com/timberio/vector/issues/3516, but for the socket source and current connections.

- [ ] Emit a `socket_connection_open_time_ms` gauge that includes the `host` as a label (might require #2660)

**Comment by ktff** [comment]
Emitting needs to happen in shared code since it handles the connections, so we have the following options:
* a) Pass types of events to emit into the shared code, then `socket_connection_open_time_ms` naming can be used.
* b) Have #2660 and emit `tcp_connection_open_time_ms` and `unix_connection_open_time_ms` instead of `socket_connection_open_time_ms`

b) is less intrusive and covers all sources that use it, but we can't use `socket_connection_open_time_ms`.

**Comment by binarylogic** [comment]
Does our approach with the `file` source (https://github.com/timberio/vector/issues/3197#issuecomment-670566583) set precedence here? Events should not have knowledge of their higher level calling context, therefore they should be specific to the call site they are implemented. So if the call site only knows about TCP and Unix, then the events should be TCP and Unix specific. Ideally, these deeper libraries would supply an API that would trigger events in a higher-level context (like we did with the `file` source), but that doesn't appear to be possible. So I'd have to lean towards `b`.

cc @lukesteensen

**Comment by ktff** [comment]
> Does our approach with the file source (#3197 (comment)) set precedence here?

Not really, that was a product of having a separate library which doesn't have access to `emit` machinery so there wasn't a choice between a) and b). But the same approach can be done in this case, it's a way of doing a). Our current approach is as you said 
> if the call site only knows about TCP and Unix, then the events should be TCP and Unix specific

I'm also leaning toward b.

**Comment by lukesteensen** [comment]
I'm not super clear on the desired result here. If we want a gauge (i.e. single value) representing how long a specific connection has been open, there are a few questions:

1. How do we differentiate between individual connections? With tags?
2. How would this data be presented on a traditional metrics dashboard? Gauges are generally a single known value and this would be a dynamic collection of values.
3. When does the gauge get updated? The desired value is "time since open", so would we just expose the open time and calculate in the presentation layer?

Basically, this is a strange use of a gauge and doesn't really seem like a metric at all.

**Comment by ktff** [comment]
> How do we differentiate between individual connections? With tags?

Yes, `host` will suffice for `tcp` connections since they are unique and guarantee that there is only one such connection open, not so sure for `unix` since it uses `inodes`. Without this guarantee, the gauge would jump between multiple values in case we ever open multiple connections with the same `host`.

> When does the gauge get updated?

A heartbeat for the connection would need to be added to be consistent with the updates. Every 1~2 sec.

> so would we just expose the open time and calculate in the presentation layer?

This is a good alternative, in which case this metric isn't needed and the presentation layer can use metrics that mark opening and closing of the connections to compute it self.

I'll hold #3583 until we resolve this, since it has similarities.

**Comment by ktff** [comment]
@binarylogic it seems to me that this gauge and the ones in #3516/#3583 aren't the way to do this as they are mostly an overhead. Overhead in a way that, with or without them anyone tracking/using this metric will need to track close events and should track open events, in case of #3516 to determine should the file be listed as open. And since open and close events are tracked, the open time can be determined. 

So I propose that we replace this and #3516 issue with ones that add missing open/close events.

If the gauges are still needed, we can add a transform or an option to `internal_metrics` source that would track these open/close events and add gauges on every our heartbeat.

/cc @lukesteensen 

Relevant comments:
https://github.com/timberio/vector/issues/3518#issuecomment-682182188
https://github.com/timberio/vector/issues/3518#issuecomment-682201356
https://github.com/timberio/vector/pull/3583#pullrequestreview-477974334
https://github.com/timberio/vector/pull/3583#issuecomment-683121684

**Comment by binarylogic** [comment]
Thanks for summarizing all of this @ktff.

> So I propose that we replace this and #3516 issue with ones that add missing open/close events.

Agree, closing this in favor of #4505.
